src/main/kotlin/org/rust/lang/core/macros/MacroExpansionTask.kt

```diff
@@ -5,578 +5,168 @@
-            } else {
-                batch.createFileWithContent(expansionBytes, stepNumber)
+    private fun collectExistingFiles(crate: CratePersistentId): Map<String, FSItem.FSFile> {
+        val path = "/$MACRO_EXPANSION_VFS_ROOT/$projectDirectoryName/$crate"
+        val rootDirectory = expansionFileSystem.getDirectory(path) ?: return emptyMap()
+        return rootDirectory
+            .copyChildren().asSequence().filterIsInstance<FSItem.FSDir>()     //  first level directories
+            .flatMap { it.copyChildren() }.filterIsInstance<FSItem.FSDir>()   // second level directories
+            .flatMap { it.copyChildren() }.filterIsInstance<FSItem.FSFile>()  // actual files
+            .associateBy {
+                val parent = it.parent
+                val grandparent = parent.parent!!
+                "${grandparent.name}${parent.name}${it.name}"
             }
-            return Stage3(info, callHash, defHash, Ok(Stage3Ok(file, ranges, expansionBytesHash, mixHash)))
-        }
     }
 
-    class Stage2OkRangesOnly(
-        private val info: ExpandedMacroInfo,
-        private val callHash: HashCode?,
-        private val defHash: HashCode?,
-        private val mixHash: HashCode,
-        private val oldExpansionFile: VirtualFile,
-        private val ranges: RangeMap,
-        private val expansionBytesHash: Long
-    ) : Pipeline.Stage2WriteToFs {
-        override fun writeExpansionToFs(batch: MacroExpansionVfsBatch, stepNumber: Int): Pipeline.Stage3SaveToStorage {
-            val file = batch.resolve(oldExpansionFile)
-            return Stage3(info, callHash, defHash, Ok(Stage3Ok(file, ranges, expansionBytesHash, mixHash)))
-        }
-    }
+    private fun deleteStaleExpansions(allDefMaps: List<CrateDefMap>): Boolean {
+        val crates = allDefMaps.mapToSet { it.crate.toString() }
 
-    class Stage2Fail(
-        private val info: ExpandedMacroInfo,
-        private val callHash: HashCode?,
-        private val defHash: HashCode?,
-        private val error: ExpansionPipelineError,
-    ) : Pipeline.Stage2WriteToFs {
-        override fun writeExpansionToFs(batch: MacroExpansionVfsBatch, stepNumber: Int): Pipeline.Stage3SaveToStorage {
-            val oldExpansionResult = info.expansionResult
-            if (oldExpansionResult is Ok && oldExpansionResult.ok.isValid) {
-                batch.deleteFile(oldExpansionResult.ok)
-            }
-            return Stage3(info, callHash, defHash, Err(error))
+        val path = "/$MACRO_EXPANSION_VFS_ROOT/$projectDirectoryName"
+        val root = expansionFileSystem.getDirectory(path) ?: return false
+        root.copyChildren()
+            .filterIsInstance<FSItem.FSFile>()
+            .filter { it.name !in crates }
+            .ifEmpty { return false }
+            .forEach { it.delete() }
+
+        expansionFileSystem.findFileByPath(path)?.let {
+            VfsUtil.markDirty(false, false, it)
```

Comment:
Could we delete, markDirty and refresh using `MacroExpansionVfsBatch` (with single `applyToVfs` per `MacroExpansionTask`)? Without it I'm not ready to assure we're doing right things here 😅. At least, I don't see the refresh (it may not be called later) and `markDirty` should likely be invoked for each deleted file.